eng/Directory.Build.Data.targets

```diff
@@ -17,11 +17,13 @@
 
   <PropertyGroup Condition="'$(IsTestProject)' == 'true' or '$(IsTestHelperLibrary)' == 'true'">
     <IsPackable>false</IsPackable>
-    <SignAssembly>false</SignAssembly>
     <RequiredTargetFrameworks>netcoreapp2.0</RequiredTargetFrameworks>
     <TargetFrameworks>$(RequiredTargetFrameworks)</TargetFrameworks>
     <DefaultReferenceTargets>AzSdk.test.reference.targets</DefaultReferenceTargets>
 
+    <!-- Disable signing of test assemblies unless it has been explicitly opted into -->
+    <SignAssembly Condition="'$(AllowTestAssemblySigning)' != 'true'">false</SignAssembly>
```

Comment:
Can you help me understand why a test assembly needs to be signed?

Reply:
If we do have a case for this we should probably use a different key from the MSSharedKey for the signing and ensure we don't get them signed by the signing service officially but only local sn signing.

Reply:
In a nutshell, I'm unable to get a clean build without doing so due to quirks in referencing  between signed and unsigned assemblies when a `friend` relationship is applied.  I wish that it weren't so, but am not aware of a way to work around it otherwise.  More detail can be found below.

If you've got a trick to loosen the reference restrictions that I overlooked or an idea for an alternate approach, I'd welcome it so that I can avoid the need to sign.   Likewise, I'd be happy to go with an alternate key if that ends up being the recommendation.   In that case, I'd ask if we would be using a common one at the engineering level or going with ones generated per-SDK where needed.

#### Scenario/Justification
- The client library makes uses of the `InternalsVisibleTo` attribute to expose non-public members to the test suite.  

- The reference to the signed client library is rejected at build time for the unsigned test library, which appears to be compliments of the `friend` designation.  

- Building with both libraries signed or both libraries unsigned works just fine.  Signing the client library and not the test suite results in a build failure.  I did not attempt to sign the test suite and link to an unsigned client library.

- The surface area of the use of internals in the existing test suite is large enough that working around it by either removing the dependency or shifting to a reflection-based approach did not seem justified, given the level of work entailed and potential risk of regressions.

- The goal of these changes is to adopt the client library into this repository with as minimal a set of changes as possible; digging in with additional analysis to validate marking the internal members `public` and understanding the surface area change to the client library did not seem to have justifiable ROI to expand the assigned task.

- While I'd agree with the statement that tests should be focused on the public contract, pragmatically, I'd also agree that it is sometimes helpful or necessary to test against members that you don't want to surface.   Use of the `InternalsVisibleTo` approach is, in my experience, reasonable and a fairly common way of doing so.   Based on that, I can see a need to support it going forward.  In my opinion, doing so via a common approach that allows hinting to the engineering system that signing is needed was a better approach of isolating local overrides on a per-SDK basis.

Reply:
For some reason I don't see my follow-up comment, not sure if github messed up or if I forget to hit post, at any rate for completeness I'm pointing out my thoughts on InternalsVisibleTo by linking it here again https://github.com/dotnet/corefx/issues/160. Although based on some of your comments it seems like you saw my earlier comment, anyways.

My dislike of InternalsVisibleTo aside I agree we should have a common mechanism for allowing tests to be signed so I suggest we generate a new test key that contains the full public/private key and commit it to the repo and use that for any test assemblies that we want to strong name sign. I'm happy to add that support or if you like you can take a shot at adding support for it. I'm also happy to do it as a follow-up to this PR and not block on it but I definitely don't want us to start signing test libraries with our MSSharedKey as a long term solution.

Reply:
I hadn't seen your earlier comment, but it does mirror my own thoughts.   :)

Per our offline discussion, I'll shift to a test-specific full key and adjust the other signing properties to accommodate.

Reply:
Changes have been made.  Signing for test assemblies is now using a newly generated key and setting properties for immediate signing.   Going to resolve, feel free to reopen if there is more discussion needed.